Installation issue: hdf (hdf4) on Ubuntu 20.04 aarch64

### Steps to reproduce the issue

Try building a container using spack with the following settings:
```
spack:
  concretization: together
  view: false

  config:
    build_jobs: 2

  compilers:
  - compiler:
      spec: gcc@10
      paths:
        cc: /usr/bin/gcc-10
        cxx: /usr/bin/g++-10
        f77: /usr/bin/gfortran-10
        fc: /usr/bin/gfortran-10
      flags: {}
      operating_system: any
      target: x86_64
      modules: []
      environment: {}
      extra_rpaths: []

  packages:
    ...

  specs:
    - hdf

  container:

    # Select the format of the recipe e.g. docker,
    # singularity or anything else that is currently supported
    format: docker
    # How to use:
    #$ spack containerize > Dockerfile
    #$ docker build -t myimage .
    #$ docker run -it myimage

    #format: singularity
    # How to use:
    #$ spack containerize > hdf5.def
    #$ sudo singularity build hdf5.sif hdf5.def

    # Sets the base images for the stages where Spack builds the
    # software or where the software gets installed after being built..
    images:
      os: ubuntu:20.04
      spack:
        url: https://github.com/spack/spack
        ref: develop
        resolve_sha: false
      
    # Whether or not to strip binaries
    strip: true

    ## Additional system packages that are needed at runtime
    os_packages:
      build:
      - gcc-10
      - g++-10
      - gfortran-10
      - cpp-10
      final:
      - gcc-10
      - g++-10
      - gfortran-10
      - cpp-10

    # Labels for the image
    labels:
      app: "test-hdf"
      mpi: "openmpi"
```


### Error message

```
#14 2970.7 ==> Installing hdf-4.2.15-rpaybquk7wffzkxv4yzyalickgpqsng2
#14 2970.7 ==> No binary for hdf-4.2.15-rpaybquk7wffzkxv4yzyalickgpqsng2 found: installing from source
#14 2971.7 ==> Fetching https://mirror.spack.io/_source-cache/archive/db/dbeeef525af7c2d01539906c28953f0fdab7dba603d1bc1ec4a5af60d002c459.tar.gz
#14 2971.7 ==> No patches needed for hdf
#14 2971.8 ==> hdf: Executing phase: 'autoreconf'
#14 2971.8 ==> hdf: Executing phase: 'configure'
#14 2978.4 ==> hdf: Executing phase: 'build'
#14 2978.8 ==> Error: ProcessError: Command exited with status 2:
#14 2978.8     'make' '-j2' 'V=1'
#14 2979.2 
#14 2979.2 2598 errors found in build log:
#14 2979.2      ...
#14 2979.2 
#14 2979.2      297     /bin/bash ../../libtool  --tag=CC   --mode=compile /opt/spack/lib/
#14 2979.2              spack/env/gcc/gcc -DHAVE_CONFIG_H -I.   -I/opt/software/linux-ubun
#14 2979.2              tu20.04-aarch64/gcc-9.4.0/zlib-1.2.12-tb6ikybgsm2t4wzyqcouo33qketg
#14 2979.2              mllh/include -I/opt/software/linux-ubuntu20.04-aarch64/gcc-9.4.0/l
#14 2979.2              ibjpeg-turbo-2.1.0-y2ewtz6dnr26i7fkpv57sjhnvd4am2p5/include   -O3 
#14 2979.2              -fomit-frame-pointer -MT bitvect.lo -MD -MP -MF .deps/bitvect.Tpo 
#14 2979.2              -c -o bitvect.lo bitvect.c
#14 2979.2      298     libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_
#14 2979.2              H -I. -I/opt/software/linux-ubuntu20.04-aarch64/gcc-9.4.0/zlib-1.2
#14 2979.2              .12-tb6ikybgsm2t4wzyqcouo33qketgmllh/include -I/opt/software/linux
#14 2979.2              -ubuntu20.04-aarch64/gcc-9.4.0/libjpeg-turbo-2.1.0-y2ewtz6dnr26i7f
#14 2979.2              kpv57sjhnvd4am2p5/include -O3 -fomit-frame-pointer -MT atom.lo -MD
#14 2979.2               -MP -MF .deps/atom.Tpo -c atom.c -o atom.o
#14 2979.2      299     libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_
#14 2979.2              H -I. -I/opt/software/linux-ubuntu20.04-aarch64/gcc-9.4.0/zlib-1.2
#14 2979.2              .12-tb6ikybgsm2t4wzyqcouo33qketgmllh/include -I/opt/software/linux
#14 2979.2              -ubuntu20.04-aarch64/gcc-9.4.0/libjpeg-turbo-2.1.0-y2ewtz6dnr26i7f
#14 2979.2              kpv57sjhnvd4am2p5/include -O3 -fomit-frame-pointer -MT bitvect.lo 
#14 2979.2              -MD -MP -MF .deps/bitvect.Tpo -c bitvect.c -o bitvect.o
#14 2979.2      300     In file included from hdf.h:20,
#14 2979.2      301                      from bitvect.h:32,
#14 2979.2      302                      from bitvect.c:103:
#14 2979.2   >> 303     hdfi.h:1119:1: error: unknown type name 'No'
#14 2979.2      304      1119 | No machine type has been defined.  Your Makefile needs to 
#14 2979.2              have someing like
#14 2979.2      305           | ^~
#14 2979.2   >> 306     hdfi.h:1119:12: error: expected '=', ',', ';', 'asm' or '__attribu
#14 2979.2              te__' before 'type'
#14 2979.2      307      1119 | No machine type has been defined.  Your Makefile needs to 
#14 2979.2              have someing like
#14 2979.2      308           |            ^~~~
#14 2979.2   >> 309     hdfi.h:1119:12: error: unknown type name 'type'; did you mean 'typ
#14 2979.2              eof'?
#14 2979.2      310      1119 | No machine type has been defined.  Your Makefile needs to 
#14 2979.2              have someing like
#14 2979.2      311           |            ^~~~
#14 2979.2      312           |            typeof
#14 2979.2      313     In file included from hdf.h:20,
#14 2979.2      314                      from atom.c:59:
#14 2979.2   >> 315     hdfi.h:1119:1: error: unknown type name 'No'
#14 2979.2      316      1119 | No machine type has been defined.  Your Makefile needs to 
#14 2979.2              have someing like
#14 2979.2      317           | ^~
#14 2979.2   >> 318     hdfi.h:1119:12: error: expected '=', ',', ';', 'asm' or '__attribu
#14 2979.2              te__' before 'type'
#14 2979.2      319      1119 | No machine type has been defined.  Your Makefile needs to 
#14 2979.2              have someing like
#14 2979.2      320           |            ^~~~
#14 2979.2      321     In file included from bitvect.h:32,
#14 2979.2      322                      from bitvect.c:103:
#14 2979.2      323     hdf.h:29:1: warning: data definition has no type or storage class
#14 2979.2      324        29 | DFdi;
#14 2979.2      325           | ^~~~
#14 2979.2      326     hdf.h:29:1: warning: type defaults to 'int' in declaration of 'DFd
#14 2979.2              i' [-Wimplicit-int]
#14 2979.2   >> 327     hdfi.h:1119:12: error: unknown type name 'type'; did you mean 'typ
#14 2979.2              eof'?
#14 2979.2      328      1119 | No machine type has been defined.  Your Makefile needs to 
#14 2979.2              have someing like
#14 2979.2      329           |            ^~~~
#14 2979.2      330           |            typeof
#14 2979.2      331     In file included from bitvect.h:32,
#14 2979.2      332                      from bitvect.c:103:
#14 2979.2   >> 333     hdf.h:142:9: error: unknown type name 'int32'
#14 2979.2      334       142 | typedef int32 HFILEID;
#14 2979.2      335           |         ^~~~~
#14 2979.2   >> 336     hdf.h:144:15: error: expected declaration specifiers or '...' befo
#14 2979.2              re '*' token
#14 2979.2      337       144 | typedef intn (*hdf_termfunc_t)(void);   /* termination fun
#14 2979.2              ction typedef */
#14 2979.2      338           |               ^
#14 2979.2      339     In file included from hdf.h:151,
#14 2979.2      340                      from bitvect.h:32,
#14 2979.2      341                      from bitvect.c:103:
#14 2979.2   >> 342     hbitio.h:35:7: error: unknown type name 'int32'
...
```

I tracked this down to Ubuntu 20.04 now being an `aarch64` system, for which hdf needs several patches according to this `nix` commit: https://github.com/knedlsepp/nixpkgs/commit/c1a2918c849a5bc766c6d55d96bc6cf85c9d27f4

### Information on your system

```
spack debug report
* **Spack:** 0.18.0.dev0 (94a9353347af4ba30c5102775ef6e3e28002e068)
* **Python:** 3.9.12
* **Platform:** darwin-monterey-x86_64
* **Concretizer:** clingo
```
But note that I was building containers, i.e. my system being macOS doesn't matter.

### Additional information

[docker.build.log.txt](https://github.com/spack/spack/files/8637470/docker.build.log.txt)

Maintainers: @lrknox 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/hdf/package.py
+++ b/var/spack/repos/builtin/packages/hdf/package.py
@@ -68,7 +68,7 @@ class Hdf(AutotoolsPackage):
     # TODO: '@:4.2.14 ~external-xdr' and the fact that we compile for 64 bit
     #  architecture should be in conflict
 
-    #https://github.com/knedlsepp/nixpkgs/commit/c1a2918c849a5bc766c6d55d96bc6cf85c9d27f4
+    # https://github.com/knedlsepp/nixpkgs/commit/c1a2918c849a5bc766c6d55d96bc6cf85c9d27f4
     patch('https://src.fedoraproject.org/rpms/hdf/raw/edbe5f49646b609f5bc9aeeee5a2be47e9556e8c/f/hdf-ppc.patch?full_index=1',
           sha256='5434f29a87856aa05124c7a9409b3ec3106c30b1ad722720773623190f6bfda8',
           when='@4.2.15:')

--- a/var/spack/repos/builtin/packages/hdf/package.py
+++ b/var/spack/repos/builtin/packages/hdf/package.py
@@ -68,6 +68,23 @@ class Hdf(AutotoolsPackage):
     # TODO: '@:4.2.14 ~external-xdr' and the fact that we compile for 64 bit
     #  architecture should be in conflict
 
+    #https://github.com/knedlsepp/nixpkgs/commit/c1a2918c849a5bc766c6d55d96bc6cf85c9d27f4
+    patch('https://src.fedoraproject.org/rpms/hdf/raw/edbe5f49646b609f5bc9aeeee5a2be47e9556e8c/f/hdf-ppc.patch?full_index=1',
+          sha256='5434f29a87856aa05124c7a9409b3ec3106c30b1ad722720773623190f6bfda8',
+          when='@4.2.15:')
+    patch('https://src.fedoraproject.org/rpms/hdf/raw/edbe5f49646b609f5bc9aeeee5a2be47e9556e8c/f/hdf-4.2.4-sparc.patch?full_index=1',
+          sha256='ce75518cccbeb80ab976b299225ea6104c3eec1ec13c09e2289913279fcf1b39',
+          when='@4.2.15:')
+    patch('https://src.fedoraproject.org/rpms/hdf/raw/edbe5f49646b609f5bc9aeeee5a2be47e9556e8c/f/hdf-s390.patch?full_index=1',
+          sha256='f7d67e8c3d0dad8bfca308936c6ac917cc0b63222c6339a29efdce14e8be6475',
+          when='@4.2.15:')
+    patch('https://src.fedoraproject.org/rpms/hdf/raw/edbe5f49646b609f5bc9aeeee5a2be47e9556e8c/f/hdf-arm.patch?full_index=1',
+          sha256='d54592df281c92e7e655b8312d18bef9ed096848de9430510e0699e98246ccd3',
+          when='@4.2.15:')
+    patch('https://src.fedoraproject.org/rpms/hdf/raw/edbe5f49646b609f5bc9aeeee5a2be47e9556e8c/f/hdf-aarch64.patch?full_index=1',
+          sha256='49733dd6143be7b30a28d386701df64a72507974274f7e4c0a9e74205510ea72',
+          when='@4.2.15:')
+
     @property
     def libs(self):
         """HDF can be queried for the following parameters:

